docs/design/fabric-v2+/fpc-key-dist.puml

```diff
@@ -280,49 +250,51 @@ group chaincode key import
 
   Admin1        -> Peer1_CLI        ++: lifecycle chaincode\n importcckeys (CC_Id)
 
-  Peer1_CLI        -> ERCC1          ++: queryECCInfos(CC_Id)
-  return ECCInfos = List<Credential, ...>
-  Peer1_CLI    -> Peer1_CLI : check if ECCInfos contains enclave for this peer;\n otherwise abort
+  Peer1_CLI        -> ERCC1          ++: queryListEnclaveCredentials(CC_Id)
+  return List<Credentials, ...> /'Admin1 -> ERCC1'/
+  Peer1_CLI    -> Peer1_CLI : check if enclave registered for this peer;\n otherwise abort
 
-  Peer1_CLI -> Peer1_CLI : Fill CC_Parameters
+  Peer1_CLI     -> ECC1             ++: importCCKeys()
+  ECC1 -> Enclave1 ++: ecall_get_enclave_id()
+  return Enclave2_Id
 
-  Peer1_CLI     -> ECC1             ++: importCCKeys(CC_Parameters)
-  ECC1          -> Enclave1         ++: importCCKeys(CC_Parameters)
-  Enclave1      -> Enclave1         :   check CC_Parameters matches internal one
-  Enclave1      -> ERCC1            ++: GetKeyDistStart(Public_Keys.Enclave2_VK)
-  return cckeydist_start_message <- <encrypted_cckeys_message, cckey_distribution_message> /'Enclave2 -> ERCC2'/
-  Enclave1      -> TLCC_Enclave1    ++: check_ercc_committed_data(cckeydist_start_message)
-  return true/false /'Enclave2 -> TLCC_Enclave2'/
-  note left
-    if false, abort
-  end note
-  Enclave1      -> Enclave1         :   extract Enclave1_VK from cckeydist_start_message.cckey_distribution_message
-  Enclave1      -> ERCC1            ++: GetRegistration(Enclave1_VK)
-  return Registered_Enclave1_Credentials /'Enclave2 -> ERCC2'/
-  Enclave1      -> TLCC_Enclave1    ++: check_ercc_committed_data(Registered_Enclave1_Credentials)
-  return true/false /'Enclave2 -> TLCC_Enclave2'/
-  note left
-    if false, abort
-  end note
-  Enclave1      -> Enclave1         :   check Enclave1_VK matches in registration_message and cckeydist_start_message
+  ECC1      -> ERCC1            ++: getKeyExport(CC_Id, Enclave2_Id)
+  return Export_message
+
+  ECC1      -> ERCC1            ++: queryEnclaveCredentials(CC_Id, Export_message.Enclave1_Id)
+  return Registered_Enclave1_Credentials
+
+  ECC1          -> Enclave1         ++: ecall_import_cc_keys(Export_message)
```

Comment:
good catch! will change

Reply:
change looks good to me but still leave unresolved just in case my security reasoning is wrong and @bvavala finds a counter-argument

Reply:
Yes it is redundant, but there may be good arguments for that, namely: it is not necessary to rely on an external party for that check.
In the non-redundant case: Enclave1 trusts whatever the TLCC-authenticated ERCC message says.
In the redundant case: Enclave1 enforces that the export message is "good".

The non-redundant case brings an interesting discussion point -- I write it here to record it, but the expectation is to discuss it elsewhere.
As TLCC arguably has a longer lifetime that chaincode enclaves, it might happen that chaincode enclaves are created following the latest platform updates, while TLCC predates the updates. So, is it reasonable that Enclave1, with a recent TCB security version number, trusts whatever comes from TLCC with an older version number?

Reply:
Right now TCB update requires a reboot, so TLCC cannot have an older TCB than ECC. And if you cannot trust TLCC you couldn't trust ERCC state so you would have to do more verification than we originally did as you also would have to validate everything happening during registration (e.g., attestation verification etc) which leads to the problem that right now only ERCC has the ground truth for IAS (that obviously could be fixed by forcing also IAS pk into ECC); also some of the checks such as consistency of tlcc mrenclave with ercc-baked in version could only approximated.

Of course it raises also questions regarding the discussion we had this morning regarding a TL-less FPC stage 1.  Probably implicit during this mornings discussion but i guess TL-less (constrained to application scenarios as discussed this morning -- single-state-object and no sensitive data release conditioned on state -- really works in current design only for designated peer but not where key distribution would come into the play. For that to work you essentially would have to replicate the ERCC functionality also inside ECC?

Reply:
> Right now TCB update requires a reboot, so TLCC cannot have an older TCB than ECC.

True, but still TLCC could have been created before the reboot (assuming it seals/unseals state, etc.), and its SVN at creation time should be reasonably exposed to Enclave1. At this point, I would imagine that the policy would be that Enclave1 refuses to establish a channel with an old TLCC and asks to rebuild it. Anyway, off-topic for now.

Reply:
Above scenario would only apply once we would make TLCC stateful so we can do faster restarts, in its current tlcc implementation/mvp plan, CPUSVN from prior incarnation wouldn't matter.  But yes, once we would add state to TLCC post-mvp, we would have to be more careful (probably would have to replace all uses of tlcc_mrencalve with the tuple <tlcc_mrenclave, min_cpusvn> in respect to what is baked into ercc and what ecc attests to and the tests would then not only have to verify identical tlcc_mrenclave but actual attested cpusvn should be >= min_cpusvn and we also would have to verify/ensure that tlcc's state is initially created and sealed with at least that min_cpusvn (similar to what we do regarding the ecc secrets)

Reply:
right, we're on the same page here.